common/rpc/interceptor/sdk_version.go

```diff
@@ -0,0 +1,97 @@
+// The MIT License
+//
+// Copyright (c) 2022 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package interceptor
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+
+	"go.temporal.io/server/common/headers"
+	"go.temporal.io/version/check"
+	"google.golang.org/grpc"
+)
+
+type SDKInfoRecorder interface {
+	RecordSDKInfo(name, version string)
+	GetAndResetSDKInfo() []check.SDKInfo
+}
+
+type SDKVersionInterceptor struct {
+	// Using a sync map to support effienct concurrent updates.
+	// Key type is sdkNameVersion and value type is sdkCount.
+	// Note that we never delete keys from this map as the cardinality of
+	// client versions should be fairly low.
```

Comment:
I wonder if we should maintain a metric of this map's size. Technically since this is a caller-provided value, one could easily blow this up from a client. There are ways we can keep the count down here if we need to.

Reply:
I'm wondering if it's worth using an LRU instead just to ensure this doesn't blow up

Reply:
We could add a counter whenever we evict so we know that something is wrong